### PR TITLE
SOF-1550 Able to TransitionEffect Actions on Next Take

### PR DESCRIPTION
**Pull request by LindvedKrvang**
- Blueprints now creates Transition Effect on Next Take Actions 
- Blueprints now creates Transition Effect Actions that are Taken immediatly
- `PieceAction` now takes an optional `InTransition` for the Part it is being inserted on

### CODE REVIEW

File: src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import { AssetFolderHelper } from '../helpers/AssetFolderHelper'
+
+const FRAME_RATE: number = 25
+const MINIMUM_DURATION_IN_MS: number = 1000
+
+export class Tv2TransitionEffectActionFactory {
+
+  constructor(
+    private readonly videoMixerTimelineObjectFactory: Tv2VideoMixerTimelineObjectFactory,
+    private readonly casparCgTimelineObjectFactory: Tv2CasparCgTimelineObjectFactory,
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts:47]: A more fleshed-out factory for CasparCG is being made in SOF-1547.
I needed to reuse some methods for CasparCG so I made this for now.
But yes, we should depend on an interface. I like your idea, that would also solve some "issues" regarding the Viz and CasparCG-implementation not having the same methods all the time.

File: src/blueprints/tv2/helpers/tv2-blueprint-configuration-mapper.ts [hunk 2]
```diff
@@ -93,4 +124,66 @@ export class Tv2BlueprintConfigurationMapper {
       }
     })
   }
+
+  private mapTransitionEffectConfigurations(transitions: string[]): TransitionEffect[] {
+    return transitions.map(transition => {
+      if (transition.match(CUT_TRANSITION_EFFECT_REGEX)) {
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/helpers/tv2-blueprint-configuration-mapper.ts:130]: But the type used for `CUT_TRANSITION_EFFECT_REGEX` is `RegExp`. That quite clearly defines it as a Regex.
I agree, it's an implementation detail. But it's an implementation detail of `Tv2BlueprintConfigurationMapper`, which is the file we are in. That's also why the variable is not exported.

File: src/business-logic/services/execute-action-service.ts [hunk 3]
```diff
@@ -69,6 +69,12 @@ export class ExecuteActionService implements ActionService {
         await this.insertPieceAsNext(pieceAction, rundownId)
         break
       }
+      case PieceActionType.INSERT_PIECE_AS_NEXT_AND_TAKE: {
+        const pieceAction: PieceAction = (await this.mutateAction(action, rundownId)) as PieceAction
+        await this.insertPieceAsNext(pieceAction, rundownId)
+        await this.rundownService.takeNext(rundownId)
+        break
+      }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/business-logic/services/execute-action-service.ts:77]: That's a boolean flag which indicates that the method has more than one responsibilities which is a bad practice.

File: src/blueprints/tv2/helpers/AssetFolderHelper.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+export class AssetFolderHelper {
+  // Copied from Blueprints
+  public joinAssetToFolder(folder: string | undefined, assetFile: string): string {
+    if (!folder) {
+      return assetFile
+    }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/helpers/AssetFolderHelper.ts:6]: As indicated on the line just above the method, this method is completely copied from Blueprints. 
I'm not a fan of this method, but I know some changes has been made to it in SOF-1547, so I'm not planning on touching it here.

File: src/blueprints/tv2/helpers/AssetFolderHelper.ts [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+    // Replace every `\\` with `\`, then replace every `\` with `/`
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/helpers/AssetFolderHelper.ts:8]: This method is completely copied from blueprints, as indicated just above the method. I know some changes are being made to it in SOF-1547 so hopefully these comments will be resolved.

File: src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+          this.audioTimelineObjectFactory.createBreakerAudioTimelineObject()
+        ]
+
+        action.data.pieceInterface.timelineObjects.push(...timelineObjects)
+        action.data.partInTransition = {
+          keepPreviousPartAliveDuration: this.getTimeFromFrames(breaker.startAlpha) + casparCgPreRollDuration,
+          delayPiecesDuration: this.getTimeFromFrames(breaker.durationInFrames - breaker.endAlpha) + casparCgPreRollDuration
+        }
+        break
+      }
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts]: I agree that the `BREAKER` case would benenfit from it. The other cases not so much as they are already calling a single function and adding the result of that function.

I have extracted the logic of the `BREAKER` case out into two methods.

File: src/blueprints/tv2/helpers/AssetFolderHelper.ts
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/helpers/AssetFolderHelper.ts:1]: Good catch. It's been updated.

File: src/blueprints/tv2/action-factories/tv2-dve-action-factory.ts [hunk 7]
```diff
@@ -519,7 +508,9 @@ export class Tv2DveActionFactory {
           isVoiceOver
         }
       },
-      data: {} as PieceInterface
+      data: {
+        pieceInterface: {} as PieceInterface
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/action-factories/tv2-dve-action-factory.ts]: I have created a `createEmptyPieceInterfaceToBeUpdatedByMutateActions()` method to be used.

File: src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            this.createBreakerTransitionEffectAction(PieceActionType.INSERT_PIECE_AS_NEXT, transitionEffect, blueprintConfiguration),
+            this.createBreakerTransitionEffectAction(PieceActionType.INSERT_PIECE_AS_NEXT_AND_TAKE, transitionEffect, blueprintConfiguration)
+          ]
+        }
+      }
+    })
+  }
+
+  public isTransitionEffectAction(action: Action): boolean {
+    const tv2Action: Tv2PartAction | Tv2PieceAction = action as Tv2PartAction | Tv2PieceAction
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts]: Updated.
Changed the parameter to `action: Tv2Action` so we don't need to cast here.

File: src/blueprints/tv2/helpers/AssetFolderHelper.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (KvelaGorrrrnio) [reply src/blueprints/tv2/helpers/AssetFolderHelper.ts]: > As indicated on the line just above the method, this method is completely copied from Blueprints.

To semi-cite yourself: We should evaluate what makes sense when moving over old code ;)

**Review by KvelaGorrrrnio — APPROVED** [review]

File: src/blueprints/tv2/timeline-object-factories/interfaces/tv2-video-mixer-timeline-object-factory.ts [hunk 9]
```diff
@@ -12,6 +13,11 @@ export interface Tv2VideoMixerTimelineObjectFactory {
   createDownstreamKeyerTimelineObject(downstreamKeyer: Tv2DownstreamKeyer, onAir: boolean): TimelineObject
   createDveBoxesTimelineObject(boxes: DveBoxProperties[], priority?: number): Tv2BlueprintTimelineObject
   createDvePropertiesTimelineObject(configuration: Tv2BlueprintConfiguration, layoutProperties: DveLayoutProperties): TimelineObject
+  createCutTransitionEffectTimelineObject(sourceInput: number): TimelineObject
+  createMixTransitionEffectTimelineObject(sourceInput: number, durationInFrames: number): TimelineObject
+  createDipTransitionEffectTimelineObject(sourceInput: number, durationInFrames: number, dipInput: number): TimelineObject
 
   getDveBoxesLayer(): string
+  getDveSourceInput(): number
+  findProgramSourceInputFromPiece(piece: Piece): number | undefined
```
[the remark this answers is not in the document]
Comment (LindvedKrvang) [reply src/blueprints/tv2/timeline-object-factories/interfaces/tv2-video-mixer-timeline-object-factory.ts:21]: Semantics? 
I don't have any strong arguments for one or the order.
The only argument I have is that, for me, `find` better indicates that it is "unable to find it" and therefore possibly returns undefined.

File: src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (andr9528) [reply src/blueprints/tv2/action-factories/tv2-transition-effect-action-factory.ts:49]: A task for creating the interfaces as suggested has been made earlier today, after a talk between me and Lindved. It is SOF-1696. Unless someone else picks it up before me, it is one i am aiming to do this sprint.

**Review by andr9528 — APPROVED** [review]

**Pull request merged**